python/paddle/reader/decorator.py

```diff
@@ -0,0 +1,60 @@
+# Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = ['buffered']
+
+from Queue import Queue
+from threading import Thread
+
+
+def buffered(reader, size):
+    """Creates a buffered data reader.
+
+    The buffered data reader will read and save data entries into a buffer.
+    Reading from the buffered data reader will proceed as long as the buffer
+    is not empty.
+    
+    Args:
+        reader: the data reader to read from.
+        size: max buffer size.
+    
+    Returns:
+        The buffered data reader.
+    """
+
+    class EndSignal():
+        pass
+
+    end = EndSignal()
+
+    def read_worker(r, q):
+        for d in r:
+            q.put(d)
+        q.put(end)
+
+    def create_reader():
```

Comment:
I checked the source code of Go's [`bufio.Reader.Read`](https://golang.org/src/bufio/bufio.go?s=4439:4489#L177). Its motivation is that it is not I/O efficient for users to call `io.Reader.Read(p []byte)` multiple times with small `len(p)`. The solution is that `bufio.Reader` reads once to fill in the big buffer `bufio.Reader.buf`, so that multiple calls of `Read` read directly from the buffer, other than from the real reader.

In our case, it seems that our reader returns only one data instance at a time, not like Go's `io.Reader.Read` which returns a sequence of multiple bytes.  So above motivation doesn't apply in our case.

I understand that you noticed a new motivation -- using asynchronous reading to accelerate reading.  However, According to http://stackoverflow.com/a/6821529/724872, it seems that Python's multi-threading doesn't improve the performance, but lowers the overall performance.

Should we abandon the idea to have a `buffered` reader decorator?  Sorry that I might have misled you towards this direction.